### ISSUE TITLE
Proposal for styled-components for `ui`

### ISSUE THREAD

**Issue by ismay** [dhis2/notes#16]
I initially wanted to type this as a reply to https://github.com/dhis2/notes/pull/14, but didn't want to derail that with a slightly unrelated discussion, so I'm posting it separately. First off, I'm not necessarily a proponent of css-in-js over css-modules (or regular css) or vice versa. I try to stay away from the holy war aspect of that debate, and just look at what works in which case.

In our case I'm seeing some problems pop up with css-modules, that I think might be solved by styled-components. Quoting from the notes:

1. down-side of isolated styles, difficult to re-use primitives of styles
2. down-side of locked down apis, no paddings/margins atm, so everything has been wrapped in another components
3. locked down to start, open up when there's a real need

So it seems to me that the goal of `ui` is to create dumb, functional components, their sole responsibility being styling (correct me if I'm wrong). A solid base for the look and feel of our apps, with the logic delegated to the apps that implement them (or d2-ui for common logic).

### Extensible

A little bit like sass's extend, with sc building on existing components can be done like so:

```javascript
// The original
const Button = styled.button`
  color: blue;
`

// extending the original
const SpecificButton = styled(Button)`
  font-weight: 600;
  margin: 0;
`
```

It allows you to use `ui` components as a base, where if you make these changes often enough you could contribute them back to the `ui` lib. Allowing this intermediary step is important I think, because when you're working on something you usually want to be able to sketch it out quickly, and deal with contributing back to upstream later, once your ideas have consolidated.

### Allows styles independent from the tag

Use whichever tag for your styles, allowing you to decouple styling headers from a specific `<h1>...<h6>` tag, or button styles from being tied to `<a>` or `<button>`:

```javascript
<Button as="a">
<Button as="button">
<Title as="h1">
<Title as="h2">
```

Which is good for accessibility, and allows the developer more flexibility, just like the extending above.

### Less bundling problems

By using css-in-js our styles can be bundled like any other react component, which in my opinion simplifies their distribution.

### Reads like regular css

I've used several css-in-js libraries, and in the end found sc the most pleasant to use. It reads like regular css, can be linted with stylelint like regular css. So you don't have to learn a new syntax since it's just css and javascript, and in my opinion it reads a lot better than style objects.

## In closing your honor,

So I think that most of my arguments here are that I think using sc will improve developer ux. In my experience it allows for a flexible ui library, that you don't have to fight and allows you to sketch ideas out quick when working on an app. A ui library will never be able to cater to all use cases, so I'm a proponent of:

> Ideological arguments in API design should always be a last choice. Unless you have a hard constraint, let the user make those choices. Otherwise you end up needlessly limiting the potential use cases you can solve for, with no upside.
> \- Abraham Lincoln

So basically trying to "make the 90% simple and the 10% possible".

There are of course a lot of comparisons out there, explaining why css-in-js is horrible, or why vanilla css is for dinosaurs. I feel that for us the balance will end up in favor of sc, but I think that we'd have to talk about this and see whether my assumptions are correct. What concerns people might have about using it, questions etc. For now, I just wanted to suggest this as my idea is that this might alleviate some of the issues we've been facing.

**Comment by ismay** [comment]
Also, I'm probably repeating a discussion that's already taken place. And my position at the moment is still quite naive as I'm quite new to the organisation. So mainly this is just something I thought I'd propose and we'll see if it's of any use, if anything it'll be a good way to document why we've chosen what we've chosen with regards to css with our SPAs.

**Comment by amcgee** [comment]
There are also [JSS](https://cssinjs.org/) which we use in most React+Mui apps today, and [styled-jsx](https://github.com/zeit/styled-jsx) which takes a slightly different approach.

**Comment by varl** [comment]
>This is an important one for me. Dealing with webpack costs a lot of time (see [here](https://twitter.com/ianstormtaylor/status/1064323419968159745)) and is not good for your cortisol levels. By using css-in-js our styles can be bundled like any other react component, which greatly simplifies their distribution.

Just for clarification: the `ui` library does not use Webpack at all*. It uses Babel for transpiling JSX/ES to es5 and CommonJS, in addition to PostCSS for preprocessing CSS. Creating bundles is left to the application.

*I lie. It is used internally by the `size-limit` package to check the total size of the library so we can catch bloat early.

**Comment by ismay** [comment]
> There are also [JSS](https://cssinjs.org/) which we use in most React+Mui apps today, and [styled-jsx](https://github.com/zeit/styled-jsx) which takes a slightly different approach.

Ah cool, seems interesting! I haven't used those libs before. Used https://github.com/paypal/glamorous and https://github.com/threepointone/glamor for a while when trying out different css in js libs.

**Comment by varl** [comment]
My experience with JSS is that it gets unreadable and messy very quickly. I grant the argument that MUI components are more complicated and more extensible than our current components but I still pose the comparison:

* MUI Button styles: https://github.com/mui-org/material-ui/blob/next/packages/material-ui/src/Button/Button.js#L12-L187
* UI Button styles: https://github.com/dhis2/ui/blob/master/src/core/Button/styles.css

They are of roughly the same size (rule-wise), but I find the second a lot easier to read and work with.

On another side, it is very simple to do logic in your JSS objects which if you are not careful can incur performance penalties. When styles are pure CSS that kind of runtime work is nicely decoupled from the styles themselves. You still have the ability to do it, but you will have to do it explicitly in your components, which I think is a fair thing to do.

`styled-components` have a better syntax for dealing with their styles inside of JS (template functions) than JSS in that sense.

**Comment by ismay** [comment]
Yeah that is a bit much (JSS version), though to be fair I do think that readability would be a bit better if the JSS version had whitespace surrounding each 'block'. Main thing I don't like about the JSS is the quotes surrounding certain keys, but that's more my personal preference.

**Comment by varl** [comment]
`styled-jsx` looks nice; it reminds me of the way VueJS does component-based styles which I found quite pleasant to work with in that framework.

**Comment by varl** [comment]
@ismay these were the guiding principles for when we started the `ui` library, putting them here for historical context, though some of those principles have proved themselves unrealistic so they aren't strictly adhered to.

## Principles

- Pure components (render based on props)
- PostCSS styled and autoprefixed components
- Avoid runtime dependencies (MUI, lodash, etc.)
- Do not use inline styles on components unless absolutely necessary
- Maximize compatibility for React versions (avoid new APIs)
- Feel should be fast, avoid animations and transitions for now.
- Use HTML elements where applicable (a button is a `<button>`, not a link element)
- Use native browser controls where possible for accessibility of e.g. screen readers, simple web browsers, low powered devices, etc.

**Comment by varl** [comment]
## Some numbers

`ui` dependencies (not the fonts) right now are rather slim, only one is needed at runtime:

- `classnames`: 371 B

If we add `styled-components` runtime:

- `styled-components`: 15.8 KB

The entire `ui` library today weighs in at 26.26 KB, adding in `styled-components` would add >50% on the bundle size. This would probably even out over time as the `ui` library grows in numbers of components.

**Comment by varl** [comment]
## Some reflections

Right now the `ui` library is based around that the application has some way to parse `import 'foobar.css'` and knows what to do with it. Typically this is accomplished through `style-loader` or `css-loader` in Webpack, so strictly speaking this library already is a CSS-in-JS library with a lot of the benefits in place as something like `styled-components` gives us.

All components have their own stylesheets which are isolated (CSS modules) from each other, so while it is not possible to delete a component+style by removing a single file, we can do it with deleting the component folder (e.g. `Button`).

Instead of using [`babel-plugin-styled-components`](https://github.com/styled-components/babel-plugin-styled-components) to precompile CSS we do it with PostCSS.

We use the same mechanisms for scoping CSS (hashes for class names).

Today we generate an RTL variation for our components using `postcss-rtl`. We would need something like that for `styled-components` as well (I haven't searched for it but I'm sure there is something).

`styled-components` would allow us to not rely on the application having a way to `import` CSS files because the runtime is bundled in the library.

The other side of the coin is that using `styled-components` in the library would impose `styled-components` on all host applications, even if that App is using Material UI or another framework.

I'm open to the idea of `styled-components` even though I'm wary of getting into the mess we had with MUI, for example:

**What happens if the lib uses `styled-components@1.0.0`, and then the Apps are using `styled-components@2.0.0`?**

**Comment by varl** [comment]
After digging in to `styled-jsx` documentation and implementation details I realized that doing a bare minimum effort port of `ui` to `styled-jsx` would be feasible.

This does a naive port of all components to `styled-jsx`: https://github.com/dhis2/ui/pull/143

It was trivial to do and already the complexity of the library feels lower. Before all built components had:

- `index.js`
- `styles.json`
- `styles.css`
- `styles.js`

And with that refactor they only need: `index.js` and `styles.js`. The CSS itself is just as complex as it was, but it can be simplified now. More information in the linked PR.

Global colors and the fonts CSS rules are not ported (those are the 4 CSS files left).

cc: @ismay @amcgee

**Comment by amcgee** [comment]
@varl nice!!  I commented on the PR as well (haven't tested it) but really like the reduced complexity and file count.  If the `styled-jsx` is functionally identical I think this is a clear win.

**Comment by varl** [comment]
@amcgee Yeah, there are a few graphical bugs because of how the CSS was structured and nested that need to be ironed out, but functionally it works just like before. "_No components were harmed in the making of this PR_".

Also gives a better way of extending the components from the host app, so I think it's aligned nicely with @ismay's comment:

>So basically trying to "make the 90% simple and the 10% possible".

**Comment by ismay** [comment]
I think I should've probably not started with suggesting a specific solution, but I should have asked first what the design goals are for the `ui` library. To understand better what problem it's meant to solve, who the intended users are, etc.

So @varl and I talked a bit about it, and I from what I understand:

-  Improve maintainability:
  - Clear separation between presentation and side-effects like fetching data. `d2-ui` is a bit hard to maintain from what I understand, partially because it combines both? So `ui` will contain only pure components
  - Wrapping material ui was also causing problems, so we're creating the components from scratch to not have to deal with that
- Improve UX: material ui contained some patterns that didn't work for us, so we're building components that do fit our needs
- Unified style
- Consumers: us and external developers. We want to have a library that's as easy to use as possible for the external devs as well

Nice btw. that the initial test helps in getting the complexity down and address some of the issues we were having, that's what I was hoping.

So I assume that whatever styling solution we use for `ui` is what we'll eventually want to use in the entire org (if it works for everyone)? Or is it something that could be decided on an app to app basis?

**Comment by varl** [comment]
>So I assume that whatever styling solution we use for `ui` is what we'll eventually want to use in the entire org (if it works for everyone)?

The idea is to roll it out to everything eventually.

>Or is it something that could be decided on an app to app basis?

This is likely how it will be to start as it is not an exhaustive set of components. So if we can keep it as light weight and as free from conflicts as possible, some apps can choose a hybrid model if that makes sense for their context.

They should keep in mind a strategy for migrating to `ui` eventually as they take that decision, but I don't see this as the most common case for core developed apps, it's more for external apps.

**Comment by ismay** [comment]
What I meant in the above comment is if we'd want to use our styling solution of choice in all our apps, or if we'd want the apps to choose their own way of doing css. @varl mentioned that that'd basically be up to each app individually. So in that case I can also imagine that going with something that would not force a certain styling library on the consuming app would be nice.

I found that material-ui basically allows you to use whatever solution you want with their components: https://material-ui.com/guides/interoperability, which might be a nice example for us (knowing that we want to move away from it as a base for our components of course).

And as to the choice of the styling library itself. I'm still quite in favor of styled-components. Personally I find it the most readable, concise syntax and in general just nice to work with. This article illustrates what I like about it fairly well: https://css-tricks.com/the-fragmented-but-evolving-state-of-css-in-js. Plus the market share of styled-components is just way beyond all the other styling libs. Which means a richer ecosystem, more active development, etc.

Granted, we're developing a ui library, and not just app components, so @varl's considerations about having to bundle or specify styled-components as a peerdep. are important points.

Interestingly, having a styled-components-like syntax and shipping components without any additional runtime is apparently possible: https://github.com/callstack/linaria.

**Comment by amcgee** [comment]
I think we should try to consolidate the styling solution used in our in-house applications, at least, so that we're not shipping multiple css-in-js runtimes in our applications and worrying about precedence.  If we want to support external app developers using different styling systems (a la MaterialUI) that might be a nice-to-have, but we should consolidate what we're using across our ecosystem if at all possible.

**Comment by varl** [comment]
@amcgee I agree that we should try to consolidate.

The most important item to me is that it is _possible_ (not optimal) to use the UI components in an app which already uses something else. Reconciling collisions and fix-ups will probably be required, but if the UI library is only usable for new apps or requires a rewrite, then wide-spread use is unfeasible in the short-to-medium term.

**Comment by amcgee** [comment]
Agree 100%

**Comment by varl** [comment]
@ismay I would not look for other solutions if I could combine straight-up CSS files with a single import for a component with CSS Modules.

Unfortunately using CSS-files with requires either the bundle software to inject a CSS loader, or to include a style injection runtime.

The code in https://github.com/dhis2/ui/pull/143 isn't perfect, but it does show the syntax and how it looks, and it is similar to `emotion` and `styled-components` in that you write CSS in template literals. Any thoughts on it?

A lot of the developer experience problems which exists today could be solved with https://github.com/dmapper/babel-plugin-react-css-modules. This would still require the bundle software to have a CSS loader however. If we decide to go that route, `styled-jsx` has the same features as well.

**Comment by varl** [comment]
Let's go with `styled-jsx` for now. :zap:
